cmd/clusterctl/client/cluster/installer.go

```diff
@@ -90,6 +109,58 @@ func installComponentsAndUpdateInventory(components repository.Components, provi
 	return providerInventory.Create(inventoryObject)
 }
 
+// waitForReady waits till the installed components are ready.
+func (i *providerInstaller) waitForReady() error {
+	log := logf.Log
+	log.Info("Waiting for providers to be available...")
+
+	if err := i.waitDeploymentsReady(); err != nil {
+		return err
+	}
+
+	// check for services too?
```

Comment:
If this could provide an additional protection from deployment being started but web hooks not yet ready to answer, I'm +1 (Probably worth to dig into controller runtime startup sequence before committing to it)

Reply:
With https://github.com/kubernetes-sigs/cluster-api/pull/4989 and https://github.com/kubernetes-sigs/controller-runtime/pull/1598 merged the deployment is considered ready only after the services are also ready.
Therefore checking to see if deployments are ready is enough. 

A quick check for this :
```shell 
❯ ./bin/clusterctl init --wait-providers=true -i docker && ./bin/clusterctl generate cluster capi-quickstart --flavor development --kubernetes-version v1.21.2 --control-plane-machine-count=1 --worker-machine-count=3 | k apply -f -
Fetching providers
Installing cert-manager Version="v1.4.0"
Waiting for cert-manager to be available...
Installing Provider="cluster-api" Version="v0.4.0" TargetNamespace="capi-system"
Installing Provider="bootstrap-kubeadm" Version="v0.4.0" TargetNamespace="capi-kubeadm-bootstrap-system"
Installing Provider="control-plane-kubeadm" Version="v0.4.0" TargetNamespace="capi-kubeadm-control-plane-system"
Installing Provider="infrastructure-docker" Version="v0.4.0" TargetNamespace="capd-system"
Waiting for providers to be available...

Your management cluster has been initialized successfully!

You can now create your first workload cluster by running the following:

  clusterctl generate cluster [name] --kubernetes-version [version] | kubectl apply -f -

cluster.cluster.x-k8s.io/capi-quickstart created
dockercluster.infrastructure.cluster.x-k8s.io/capi-quickstart created
dockermachinetemplate.infrastructure.cluster.x-k8s.io/capi-quickstart-control-plane created
kubeadmcontrolplane.controlplane.cluster.x-k8s.io/capi-quickstart-control-plane created
dockermachinetemplate.infrastructure.cluster.x-k8s.io/capi-quickstart-md-0 created
kubeadmconfigtemplate.bootstrap.cluster.x-k8s.io/capi-quickstart-md-0 created
machinedeployment.cluster.x-k8s.io/capi-quickstart-md-0 created
```
shows that waiting for deployments is a good enough check.

Reply:
@ykakarap I agree that taking a look at the service won't significantly improve it. But actually the sequence should be something like:

1. CAPI controllers are deployed
2. reconciler and webhooks in CAPI controllers start
3. Pod and thus Deployments are ready after webhooks answers a net.DialTimeout successfully (i.e. they are up)
4. The kube-controller-manager adds the Pod IP to the endpoints of the service
5. kube-proxy reconciles the service (e.g. adjusts iptables to handle the new service IP => pod IP mapping correctly)

So after this PR we wait until after 3. By looking at the service we could also make sure 4 has completed, but 5 is only safe after either:
* pinging the webhook (which is imho not that easy to implement in clusterctl)
* (probably) waiting a few seconds after the Pod is ready

Imho it's fine to not add additional handling to this PR. One easy option we have if we still get user reports with this error is to add `minReadySeconds` in our controller deployments (which would have to be done by all providers)

Reply:
Thanks @sbueringer, for that detailed sequence of operations.
Very informational 😃

Reply:
Who know, when we'll need it :) (hopefully never)

Reply:
Another thank-you for that sequence @sbueringer!

I've hit this issue regularly. I _do_ want `clusterctl` to tell me whether the controllers/webhooks have "deployed successfully," because that tells me that it's at least _possible_ for webhook requests to succeed. But that does not _guarantee_ that webhooks requests will succeed. So clients should still be built with the assumption that webhook requests may not succeed, or that kube-apiserver may suffer some other "internal error."

Reply:
Yup, I guess the only really 100% reliable solution is to retry the apply in case the webhooks are either not yet or currently not available.

Reply:
Agreed. Internal server errors happen, and users must be prepared. But users shouldn't be waiting for controllers or webhooks about which they know nothing. That's where this PR is useful: it helps clusterctl tell users whether the controllers/webhooks deployed successfully once, even if for a moment.